# Explorando y validando el dataset de los olimpicos

Importando librerìas

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


Leemos el dataset

In [2]:
df = pd.read_csv('1 - Olympics_without_validating.csv')
df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  № Summer  01 !  02 !  03 !  Total  № Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  № Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

Con el metodo .info() se muestra:
- Cantidad de finals.
- Cantidad de columnas.
- Cada columna si posee datos faltantes.
- El tipo de dato de cada columna

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148 entries, 0 to 147
Data columns (total 16 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       147 non-null    object
 1   1       148 non-null    object
 2   2       148 non-null    object
 3   3       148 non-null    object
 4   4       148 non-null    object
 5   5       148 non-null    object
 6   6       148 non-null    object
 7   7       148 non-null    object
 8   8       148 non-null    object
 9   9       148 non-null    object
 10  10      148 non-null    object
 11  11      148 non-null    object
 12  12      148 non-null    object
 13  13      148 non-null    object
 14  14      148 non-null    object
 15  15      148 non-null    object
dtypes: object(16)
memory usage: 18.6+ KB


Vemos el nombre de cada columna del dataframe

In [4]:
df.columns

Index(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',
       '13', '14', '15'],
      dtype='object')

Se crea un dataframe prueba a partir del dataframe creado por el dataset importado, para realizar cambios y evitar dañar el dataset.

In [5]:
dfprueba = pd.DataFrame(df)

Como se ve en el dataframe, la fila numero 0 contiene el nombre de las columnas pero se muestra como un dato mas del dataframe

In [6]:
dfprueba.loc[0]

0                NaN
1           № Summer
2               01 !
3               02 !
4               03 !
5              Total
6           № Winter
7               01 !
8               02 !
9               03 !
10             Total
11           № Games
12              01 !
13              02 !
14              03 !
15    Combined total
Name: 0, dtype: object

Por lo anterior, dicha fila que contiene los titulos de las columnas se establecen como las columnas del dataframe.

In [7]:
dfprueba.columns = dfprueba.loc[0]

Al momento de establecer dicha fila como los titulos de las columnas, procedemos a eliminarla.

In [8]:
dfprueba.drop(0, axis=0, inplace=True)

Con el metodo .reset_index() establecemos de nuevo el dataset ya que al borrar la fila 0, el dataset empieza desde el indice 1, lo reestablecemos para empezar de 0 de nuevo.

In [9]:
dfprueba.reset_index(inplace=True)

Al reestablecer el indice, se crea una nueva columna llamada index (la cual empieza desde el indice 1), por lo que se procede a eliminarla.

In [10]:
dfprueba.drop('index', axis=1, inplace=True)

Se establecen nuevos nombres para las columnas y asi poder entender mejor el dataframe.

In [15]:
dfprueba.columns = ['Pais','N_Summer','Gold_Summer','Silver_Summer','Bronze_Summer','Total_Summer','N_Winter','Gold_Winter','Silver_Winter','Bronze_Winter','Total_Winter','N_Games','Gold_Total','Silver_Total','Bronze_Total','Combined total']

In [16]:
dfprueba.head()

Pais N_Summer Gold_Summer Silver_Summer Bronze_Summer  \
0        Afghanistan (AFG)       13           0             0             2   
1            Algeria (ALG)       12           5             2             8   
2          Argentina (ARG)       23          18            24            28   
3            Armenia (ARM)        5           1             2             9   
4  Australasia (ANZ) [ANZ]        2           3             4             5   

  Total_Summer N_Winter Gold_Winter Silver_Winter Bronze_Winter Total_Winter  \
0            2        0           0             0             0            0   
1           15        3           0             0             0            0   
2           70       18           0             0             0            0   
3           12        6           0             0             0            0   
4           12        0           0             0             0            0   

  N_Games Gold_Total Silver_Total Bronze_Total Combined total  
0      13          0            0            2              2  
1      15          5            2            8             15  
2      41         18           24           28             70  
3      11          1            2            9             12  
4       2          3            4            5             12

Se observa la columna 'Pais' la cual se pretende separar el nombre del pais y su codigo.

In [17]:
dfprueba['Pais']

0                                 Afghanistan (AFG)
1                                     Algeria (ALG)
2                                   Argentina (ARG)
3                                     Armenia (ARM)
4                           Australasia (ANZ) [ANZ]
                           ...                     
142    Independent Olympic Participants (IOP) [IOP]
143                              Zambia (ZAM) [ZAM]
144                            Zimbabwe (ZIM) [ZIM]
145                          Mixed team (ZZX) [ZZX]
146                                          Totals
Name: Pais, Length: 147, dtype: object

Se crea una lista 'Codigo' la cual contiene una lista de los paises y el codigo divididos

In [18]:
Codigo = []
Codigo= dfprueba['Pais'].str.split('(')

In [19]:
Codigo= list(Codigo)


Se elimina el ultimo dato de la lista 'Codigo'.

In [20]:
Codigo.pop()


['Totals']

Y se agrega '(' un parentesis por cuestion de mejor visualizacion y se agrega a una lista nueva llamada 'codigo'.

In [21]:
codigo= []
for cod in Codigo:
    cod1 = ('('+ cod[1])
    codigo.append(cod1)

Finalmente se agrega la palabra 'WORLD' para igualar la cantidad de filas que contiene la lista y el dataframe y se crea la columna 'Codigo'.

In [22]:
codigo.append('WORLD')
dfprueba['Codigo'] = codigo 

In [23]:
dfprueba

Pais N_Summer Gold_Summer  \
0                               Afghanistan (AFG)       13           0   
1                                   Algeria (ALG)       12           5   
2                                 Argentina (ARG)       23          18   
3                                   Armenia (ARM)        5           1   
4                         Australasia (ANZ) [ANZ]        2           3   
..                                            ...      ...         ...   
142  Independent Olympic Participants (IOP) [IOP]        1           0   
143                            Zambia (ZAM) [ZAM]       12           0   
144                          Zimbabwe (ZIM) [ZIM]       12           3   
145                        Mixed team (ZZX) [ZZX]        3           8   
146                                        Totals       27        4809   

    Silver_Summer Bronze_Summer Total_Summer N_Winter Gold_Winter  \
0               0             2            2        0           0   
1               2             8           15        3           0   
2              24            28           70       18           0   
3               2             9           12        6           0   
4               4             5           12        0           0   
..            ...           ...          ...      ...         ...   
142             1             2            3        0           0   
143             1             1            2        0           0   
144             4             1            8        1           0   
145             5             4           17        0           0   
146          4775          5130        14714       22         959   

    Silver_Winter Bronze_Winter Total_Winter N_Games Gold_Total Silver_Total  \
0               0             0            0      13          0            0   
1               0             0            0      15          5            2   
2               0             0            0      41         18           24   
3               0             0            0      11          1            2   
4               0             0            0       2          3            4   
..            ...           ...          ...     ...        ...          ...   
142             0             0            0       1          0            1   
143             0             0            0      12          0            1   
144             0             0            0      13          3            4   
145             0             0            0       3          8            5   
146           958           948         2865      49       5768         5733   

    Bronze_Total Combined total       Codigo  
0              2              2        (AFG)  
1              8             15        (ALG)  
2             28             70        (ARG)  
3              9             12        (ARM)  
4              5             12  (ANZ) [ANZ]  
..           ...            ...          ...  
142            2              3  (IOP) [IOP]  
143            1              2  (ZAM) [ZAM]  
144            1              8  (ZIM) [ZIM]  
145            4             17  (ZZX) [ZZX]  
146         6078          17579        WORLD  

[147 rows x 17 columns]

Ahora para actualizar la columna pais y que muestre solo el nombre del pais se recorre la lista Codigo y se agrega cada valor (nombre del pais - sin codigo) a una nueva lista 'pais1'.

In [24]:
pais1= []
for pais in Codigo:
    pais1.append(pais[0])

Se agrega la palabra 'Totals' para igualar el numero de datos que contiene la lista y el dataframe.

In [25]:
pais1.append('Totals')

Y se actualiza la columna pais, con los nombres de cada uno de los paises.

In [26]:
dfprueba['Pais']= pais1

In [27]:
dfprueba

Pais N_Summer Gold_Summer Silver_Summer  \
0                         Afghanistan        13           0             0   
1                             Algeria        12           5             2   
2                           Argentina        23          18            24   
3                             Armenia         5           1             2   
4                         Australasia         2           3             4   
..                                 ...      ...         ...           ...   
142  Independent Olympic Participants         1           0             1   
143                            Zambia        12           0             1   
144                          Zimbabwe        12           3             4   
145                        Mixed team         3           8             5   
146                             Totals       27        4809          4775   

    Bronze_Summer Total_Summer N_Winter Gold_Winter Silver_Winter  \
0               2            2        0           0             0   
1               8           15        3           0             0   
2              28           70       18           0             0   
3               9           12        6           0             0   
4               5           12        0           0             0   
..            ...          ...      ...         ...           ...   
142             2            3        0           0             0   
143             1            2        0           0             0   
144             1            8        1           0             0   
145             4           17        0           0             0   
146          5130        14714       22         959           958   

    Bronze_Winter Total_Winter N_Games Gold_Total Silver_Total Bronze_Total  \
0               0            0      13          0            0            2   
1               0            0      15          5            2            8   
2               0            0      41         18           24           28   
3               0            0      11          1            2            9   
4               0            0       2          3            4            5   
..            ...          ...     ...        ...          ...          ...   
142             0            0       1          0            1            2   
143             0            0      12          0            1            1   
144             0            0      13          3            4            1   
145             0            0       3          8            5            4   
146           948         2865      49       5768         5733         6078   

    Combined total       Codigo  
0                2        (AFG)  
1               15        (ALG)  
2               70        (ARG)  
3               12        (ARM)  
4               12  (ANZ) [ANZ]  
..             ...          ...  
142              3  (IOP) [IOP]  
143              2  (ZAM) [ZAM]  
144              8  (ZIM) [ZIM]  
145             17  (ZZX) [ZZX]  
146          17579        WORLD  

[147 rows x 17 columns]

Ahora reordenamos las columnas del dataframe de la siguiente manera:

In [28]:
dfprueba = dfprueba[['Pais','Codigo', 'N_Summer', 'Gold_Summer', 'Silver_Summer', 'Bronze_Summer',
       'Total_Summer', 'N_Winter', 'Gold_Winter', 'Silver_Winter',
       'Bronze_Winter', 'Total_Winter', 'N_Games', 'Gold_Total',
       'Silver_Total', 'Bronze_Total', 'Combined total']]

Y cambiamos el tipo de datos de las columnas que contienen numeros.

In [29]:
dfprueba[['N_Summer', 'Gold_Summer', 'Silver_Summer', 'Bronze_Summer',
       'Total_Summer', 'N_Winter', 'Gold_Winter', 'Silver_Winter',
       'Bronze_Winter', 'Total_Winter', 'N_Games', 'Gold_Total',
       'Silver_Total', 'Bronze_Total', 'Combined total']] = dfprueba[['N_Summer', 'Gold_Summer', 'Silver_Summer', 'Bronze_Summer',
       'Total_Summer', 'N_Winter', 'Gold_Winter', 'Silver_Winter',
       'Bronze_Winter', 'Total_Winter', 'N_Games', 'Gold_Total',
       'Silver_Total', 'Bronze_Total', 'Combined total']].astype(int)

Corroboramos y efectivamente cambio el tipo de datos de las columnas que contiene datos numericos. Inicialmente todas las columnas del dataframe eran de tipo object.

In [30]:
dfprueba.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 147 entries, 0 to 146
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Pais            147 non-null    object
 1   Codigo          147 non-null    object
 2   N_Summer        147 non-null    int32 
 3   Gold_Summer     147 non-null    int32 
 4   Silver_Summer   147 non-null    int32 
 5   Bronze_Summer   147 non-null    int32 
 6   Total_Summer    147 non-null    int32 
 7   N_Winter        147 non-null    int32 
 8   Gold_Winter     147 non-null    int32 
 9   Silver_Winter   147 non-null    int32 
 10  Bronze_Winter   147 non-null    int32 
 11  Total_Winter    147 non-null    int32 
 12  N_Games         147 non-null    int32 
 13  Gold_Total      147 non-null    int32 
 14  Silver_Total    147 non-null    int32 
 15  Bronze_Total    147 non-null    int32 
 16  Combined total  147 non-null    int32 
dtypes: int32(15), object(2)
memory usage: 11.0+ KB


Con el cambio en el tipo de datos procedemos a observar una descripcion estadistica del dataset donde muestra los siguientes datos en cada columna:
- Numero de datos.
- El promedio.
- La desviacion estandar.
- El valor minimo.
- El valor del percentil 25.
- El valor del percentil 50 (mediana).
- El valor del percentil 75.
- El valor maximo.

In [31]:
dfprueba.loc[0:145].describe()

N_Summer  Gold_Summer  Silver_Summer  Bronze_Summer  Total_Summer  \
count  146.000000   146.000000     146.000000     146.000000    146.000000   
mean    13.383562    32.938356      32.705479      35.136986    100.780822   
std      7.006657    96.745184      80.648421      77.118416    252.700436   
min      1.000000     0.000000       0.000000       0.000000      0.000000   
25%      8.000000     0.000000       1.000000       1.000000      2.000000   
50%     13.000000     3.000000       4.000000       6.000000     12.000000   
75%     18.000000    23.000000      26.750000      28.750000     85.250000   
max     27.000000   976.000000     757.000000     666.000000   2399.000000   

         N_Winter  Gold_Winter  Silver_Winter  Bronze_Winter  Total_Winter  \
count  146.000000   146.000000     146.000000     146.000000    146.000000   
mean     6.595890     6.568493       6.561644       6.493151     19.623288   
std      7.349012    18.975086      18.606996      17.499654     54.640914   
min      0.000000     0.000000       0.000000       0.000000      0.000000   
25%      0.000000     0.000000       0.000000       0.000000      0.000000   
50%      5.000000     0.000000       0.000000       0.000000      0.000000   
75%     10.000000     0.750000       1.750000       1.000000      4.750000   
max     22.000000   118.000000     111.000000     100.000000    329.000000   

          N_Games   Gold_Total  Silver_Total  Bronze_Total  Combined total  
count  146.000000   146.000000    146.000000    146.000000      146.000000  
mean    19.979452    39.506849     39.267123     41.630137      120.404110  
std     13.084062   109.932471     93.984393     89.826117      291.988913  
min      1.000000     0.000000      0.000000      0.000000        1.000000  
25%     11.000000     0.000000      1.000000      1.000000        2.250000  
50%     15.000000     3.000000      4.000000      6.500000       12.000000  
75%     26.000000    24.500000     28.000000     29.000000       87.750000  
max     49.000000  1072.000000    859.000000    750.000000     2681.000000

# VALIDACIONES

Con los cambios realizados procedemos a realizar validaciones para verificar la veracidad del dataset.

In [32]:
dfprueba.columns

Index(['Pais', 'Codigo', 'N_Summer', 'Gold_Summer', 'Silver_Summer',
       'Bronze_Summer', 'Total_Summer', 'N_Winter', 'Gold_Winter',
       'Silver_Winter', 'Bronze_Winter', 'Total_Winter', 'N_Games',
       'Gold_Total', 'Silver_Total', 'Bronze_Total', 'Combined total'],
      dtype='object')

Validacion del numero de juegos:

La siguiente validacion muestra que la suma del nùmero de juegos realizados en el verano y el nùmero de juegos realizados en inviernos debe ser igual al nùmero total de juegos en los olimpicos para cada uno de los paises.

In [33]:
Validacion_N_Juegos = dfprueba['N_Summer'] + dfprueba['N_Winter'] == dfprueba['N_Games']

La variable 'Validacion_N_Juegos' muesta una tabla con valores True o False que determinan si es correcta o no la condicion. Se realiza la suma de los valores y debe ser igual a 147, ya que es el nùmero de filas que contiene el dataset.

In [34]:
Validacion_N_Juegos.sum()

147

Validacion de medallas obtenidas en verano para cada pais.

In [35]:
Validacion_Medals_Summer = dfprueba[['Gold_Summer', 'Silver_Summer',
       'Bronze_Summer']].sum(axis=1) == dfprueba['Total_Summer']

Validacion_Medals_Summer.sum()

147

Validacion de medallas obtenidas en invierno para cada pais.

In [36]:
Validacion_Medals_Winter = dfprueba[['Gold_Winter', 'Silver_Winter',
       'Bronze_Winter']].sum(axis=1) == dfprueba['Total_Winter']

Validacion_Medals_Winter.sum()

147

Validacion del total de medallas para cada pais.

In [37]:
Validacion_Total_Medals = dfprueba['Total_Summer'] + dfprueba['Total_Winter'] == dfprueba['Combined total']
Validacion_Total_Medals.sum()

147

Validacion de medallas de oro obtenidas para cada pais.

In [38]:
Validacion_Total_Golds = dfprueba['Gold_Summer'] + dfprueba['Gold_Winter'] == dfprueba['Gold_Total']
Validacion_Total_Golds.sum()

147

Validacion de medallas de plata obtenidas para cada pais.

In [39]:
Validacion_Total_Silver = dfprueba['Silver_Summer'] + dfprueba['Silver_Winter'] == dfprueba['Silver_Total']
Validacion_Total_Silver.sum()

147

Validacion de medallas de bronce obtenidas para cada pais.

In [40]:
Validacion_Total_Bronze = dfprueba['Bronze_Summer'] + dfprueba['Bronze_Winter'] == dfprueba['Bronze_Total']
Validacion_Total_Bronze.sum()

147

Validacion de todas las columnas que representan medallas:
Aqui se realiza la suma todas las filas por cada columna y se compara con la ultima fila que representa el total de medallas de cada columna.

In [41]:
Validacion_Medals_Totals = dfprueba[['Gold_Summer', 'Silver_Summer', 'Bronze_Summer',
       'Total_Summer', 'Gold_Winter', 'Silver_Winter',
       'Bronze_Winter', 'Total_Winter', 'Gold_Total',
       'Silver_Total', 'Bronze_Total', 'Combined total']][0:146].sum() == dfprueba[['Gold_Summer', 'Silver_Summer', 'Bronze_Summer',
       'Total_Summer', 'Gold_Winter', 'Silver_Winter',
       'Bronze_Winter', 'Total_Winter','Gold_Total',
       'Silver_Total', 'Bronze_Total', 'Combined total']].loc[146]

Como vemos la validaciones es correcta y se cumple la condicion.

In [42]:
Validacion_Medals_Totals

Gold_Summer       True
Silver_Summer     True
Bronze_Summer     True
Total_Summer      True
Gold_Winter       True
Silver_Winter     True
Bronze_Winter     True
Total_Winter      True
Gold_Total        True
Silver_Total      True
Bronze_Total      True
Combined total    True
dtype: bool

Con todo lo realizado anterior vemos que el dataset es veridico y se exporta para poder realizar un proximo analisis de los datos.

In [53]:
dfprueba.to_csv('3 - Real_Olympics_Dataset.csv')